raiden/network/proxies/token_network.py

```diff
@@ -959,6 +1007,18 @@ def settle(
                     locked_amount,
                     locksroot,
                 )
+                if not gas_limit:
+                    # Context switch before we check the channel data in
+                    # case the partner already settled and this is why settle would fail
+                    gevent.sleep(1)
```

Comment:
Do we need a 1sec wait here, or is just a context switch enough?

Reply:
> but had not updated his view of the channel by processing the event

The proxy does not access Raiden's internal state, this can not be the reason.

What was the exception? Because I think the race is still there, you just have to have a difference of 1s among the calls to settle.

Reply:
> What was the exception? 

The exception was `raise RaidenUnrecoverableError('Settle transaction will always fail')` which happens at the end of this `if` block. As I said above It just meant that `estimateGas()` sees the transaction will fail, returns `None` and then it's probably the race we said.

> The proxy does not access Raiden's internal state, this can not be the reason.

Yes indeed. This is not the state I meant. And thinking about it context switching is not needed -- it's the waiting that worked.

What I meant is the onchain state by calling the `detail_channel()` function via the [_check_channel_state_for_settle](https://github.com/raiden-network/raiden/blob/761bedfee2ee326401ad5ec95d55b1ab458a5213/raiden/network/proxies/token_network.py#L1151) call. 

- The 1 second delay allows for the block containing the partner's settle to be mined for the test.
- Then `detail_channel()` returns that the channel state is settled.
- And knowing this we can say that `estimateGas` failed since the channel is already settled.

But writing all these out I am sure that the 1 second of waiting there is not a good solution. It made the test pass but that's not guaranteeing that the race won't happen in a real scenario with longer block times. A different solution is needed.